File: baml_language/sdk_tests/crates/swift/type_shapes/customizable/roundtrip_tests/TestForwardRefs.swift
// Roundtrip coverage for `Baml.forward_refs` — port of python_pydantic2
// `roundtrip_tests/test_forward_refs.py`.
//
// Not yet ported: g_node_int (generic classes, Phase 5).
// `Node` (required self-ref) emits and compiles — boxed — but is
// uninhabitable from the host side, exactly like Python's import-only
// treatment.
import XCTest
import Baml

final class TestForwardRefs: XCTestCase {
    func test_forward_refs_round_trip_rec_list() throws {
        // Python value [1, [2, 3]].
        let r: Baml.forward_refs.RecList = .t1([.t0(1), .t1([.t0(2), .t0(3)])])
        XCTAssertEqual(try Baml.forward_refs.round_trip_rec_list(r: r), r)
    }

    func test_forward_refs_round_trip_rec_list_with_other() throws {
        // RecListWithOther = int | Other | RecListWithOther[].
        XCTAssertEqual(
            try Baml.forward_refs.round_trip_rec_list_with_other(r: .t0(1)),
            .t0(1)
        )
        let listy: Baml.forward_refs.RecListWithOther = .t2([.t0(1), .t0(2)])
        XCTAssertEqual(try Baml.forward_refs.round_trip_rec_list_with_other(r: listy), listy)
    }

    func test_forward_refs_round_trip_other() throws {
        let o = Baml.forward_refs.Other(v: 7)
        XCTAssertEqual(try Baml.forward_refs.round_trip_other(o: o), o)
    }

    func test_forward_refs_round_trip_node_symbol_exists() {
        // Compile-time analog of Python's import-only assertion for the
        // uninhabitable `Node` (required self-reference).
        _ = Baml.forward_refs.Node.self
    }
}

File: baml_language/sdk_tests/crates/java/type_shapes/customizable/roundtrip_tests/TestForwardRefs.java
// Roundtrip coverage for `baml_sdk.forward_refs` — forward references.
//
// Port of python_pydantic2/type_shapes/customizable/roundtrip_tests/
// test_forward_refs.py — same test names, cases, inputs, assertions.
//
// java-port note: `round_trip_node` is intentionally NOT exercised by a
// `@Test` — `class Node { next Node }` has a *required* (non-optional)
// self-reference, so no finite value can be constructed from the host side.
// Python proves the symbol is merely reachable via a module-level
// `# noqa: F401` import; the closest Java analog is a method-reference
// field that forces javac to resolve `Fns.round_trip_node`'s signature
// without ever calling it.
package roundtrip_tests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import baml_sdk.forward_refs.Fns;
import baml_sdk.forward_refs.GNode;
import baml_sdk.forward_refs.Node;
import baml_sdk.forward_refs.Other;
import baml_sdk.forward_refs.RecList;
import baml_sdk.forward_refs.RecListWithOther;
import java.util.List;
import java.util.function.UnaryOperator;
import org.junit.jupiter.api.Test;

class TestForwardRefs {

    @SuppressWarnings("unused")
    private static final UnaryOperator<Node> ROUND_TRIP_NODE_REACHABILITY_PROBE =
            Fns::round_trip_node;

    @Test
    void test_forward_refs_round_trip_other() {
        Other o = new Other(7L);
        assertEquals(o, Fns.round_trip_other(o));
    }

    @Test
    void test_forward_refs_round_trip_rec_list() {
        RecList r =
                new RecList.RecListListValue(
                        List.of(
                                new RecList.IntValue(1L),
                                new RecList.RecListListValue(
                                        List.of(
                                                new RecList.IntValue(2L),
                                                new RecList.IntValue(3L)))));
        assertEquals(r, Fns.round_trip_rec_list(r));
    }

    @Test
    void test_forward_refs_round_trip_rec_list_with_other() {
        // RecListWithOther = int | Other | RecListWithOther[]
        assertEquals(
                new RecListWithOther.IntValue(1L),
                Fns.round_trip_rec_list_with_other(new RecListWithOther.IntValue(1L)));

        RecListWithOther r =
                new RecListWithOther.RecListWithOtherListValue(
                        List.of(
                                new RecListWithOther.IntValue(1L),
                                new RecListWithOther.IntValue(2L)));
        assertEquals(r, Fns.round_trip_rec_list_with_other(r));
    }

    @Test
    void test_forward_refs_round_trip_g_node_int() {
        // The leaf node carries `children=[]`; this exercises the empty-list
        // round trip fixed under Bug A (35b). See TestGenerics.java for the
        // java-port note on the `new GNode<Long>(...)` generic-constructor
        // shape this port assumes.
        GNode<Long> g = new GNode<>(List.of(new GNode<>(List.of())));
        assertEquals(g, Fns.round_trip_g_node_int(g));
    }
}
